### DISCUSSION TITLE
Social image meta tag

### DISCUSSION THREAD

**Discussion by werls** [11ty/eleventy#2439] [Q&A] [answered]
Hi! I'm looking for a way to define page's social-image meta tag based on the image linked on the first img tag of the html's main section. As I'm using the eleventy-img plugin for all images of my site, this seems to be an easy way to solve this.

In isolation, this piece of code would solve the problem: 

```
const main = document.querySelector('main')
const firstImgTag = main.querySelector('img')

let metaImg = document.createElement('meta')
metaImg.setAttribute('property', 'og:image')
metaImg.setAttribute('content', firstImgTag.src)

document.head.append(metaImg)
```

I believe that this code (or another of the same type) should be applied after image conversions and with the html already formed by the build process. The result I expect is that the `<meta>` tag is already present in the final version of the build, as part of the static site.

I'm not very familiar with Node.js and the Eleventy build process, so I'm wondering what's the best way to solve this process. I appreciate any suggestions.

**Comment by pdehaan** [comment, marked as answer]
I'd start by looking at https://www.11ty.dev/docs/config/#transforms.
Transforms run pretty late in the build pipeline when the HTML should be "complete". 
Then, check whether the current file being transformed is an .html template, and if so, use something like https://cheerio.js.org/ to query the markup for your first `main img` tag and then append your `<meta property="og:image" content="{{ something }}"/>` tag into the `<head>` and then return the new HTML DOM which will be written to the disk.

**Reply by Ryuno-Ki** [reply]
Handle the case where a HTML might not have an image as well (perhaps you have a default image?)

**Reply by pdehaan** [reply]
I was also going to say handle the case of where you have a _specific_ image you might want to use (instead of always just the first `<img>` tag); but now I'm just thinking it might be easier to just specify the `og:image` name (which should really be fully qualified w/ domain name, etc) in the front matter and then you don't even need a transformer and you can just handle the `og:image` tag in your base layout. A bit more duplication, but faster build times and fewer moving parts.

**Reply by werls** [reply]
> I'd start by looking at https://www.11ty.dev/docs/config/#transforms. Transforms run pretty late in the build pipeline when the HTML should be "complete". Then, check whether the current file being transformed is an .html template, and if so, use something like https://cheerio.js.org/ to query the markup for your first `main img` tag and then append your `<meta property="og:image" content="{{ something }}"/>` tag into the `<head>` and then return the new HTML DOM which will be written to the disk.

Thank you, @pdehaan! I was thinking about the Transforms too, but I haven't figure out how to change the html content. I'll read more about cheerio.js and do some testing.

> Handle the case where a HTML might not have an image as well (perhaps you have a default image?)

Thanks, @Ryuno-Ki! This website in question is a design portfolio, so all the pages will have one image at least. But sure this is a nice complement to a fully functional code.

> I was also going to say handle the case of where you have a _specific_ image you might want to use (instead of always just the first `<img>` tag); but now I'm just thinking it might be easier to just specify the `og:image` name (which should really be fully qualified w/ domain name, etc) in the front matter and then you don't even need a transformer and you can just handle the `og:image` tag in your base layout. A bit more duplication, but faster build times and fewer moving parts.

That's my first idea but, as I'm using the eleventy-img to convert the images, I haven't figure out a way to manage this. Thinking again, the best way I thought was to set the filename (without the extension) in the frontmatter of the .md file of the post. Since the file is already being processed through eleventy-img, I should be able to set the meta tag directly, something like `<meta name="og:image" content="https://my.website.url{{page.url}}{{metaImg}}-400w.jpeg">`.

Does this make sense with your idea or did you think of another way, @pdehaan?

**Reply by Ryuno-Ki** [reply]
> I haven't figure out how to change the html content

Untested, but I'd assume something along these lines:

```js
eleventyConfig.addTransform("og-image", function(content) {
  if (outputPath && outputPath.endsWith(".html")) {
    const $ = cheerio.load(markup);
    const $firstImage = $('main img');
    $('head').append(`<meta name="og:image" content="${$firstImage.attr('src')}" />`);
    return $.html();
  }

  // Don't modify otherwise
  return content;
})
```

**Reply by werls** [reply]
Thank you @Ryuno-Ki, this is exactly what I was looking for! Just needed some tweaking: `... function(content, outputPath) { ...` and `const $ = cheerio.load(content);`. The working version of the code:

```javascript
eleventyConfig.addTransform("og-image", function(content, outputPath) {
  if (outputPath && outputPath.endsWith(".html")) {
    const $ = cheerio.load(content);
    const $firstImage = $('main img');
    $('head').append(`<meta name="og:image" content="https://my.website.url${$firstImage.attr('src')}">`);
    return $.html();
  }

  // Don't modify otherwise
  return content;
})
```

> That's my first idea but, as I'm using the eleventy-img to convert the images, I haven't figure out a way to manage this. Thinking again, the best way I thought was to set the filename (without the extension) in the frontmatter of the .md file of the post. Since the file is already being processed through eleventy-img, I should be able to set the meta tag directly, something like <meta name="og:image" content="https://my.website.url{{page.url}}{{metaImg}}-400w.jpeg">.

I also tested this idea based on @pdehaan solution and it works perfectly, although it needs more work to change the front matter of all the pages, among other adjustments that proved necessary in my case (gif images, for example, which have a specific treatment). Despite that, in situations where the build time difference is significant, this seems to be a good solution.

Thank you all!

**Reply by Ryuno-Ki** [reply]
:partying_face: 

Would you mind to [mark a comment as an answer](https://docs.github.com/en/discussions/collaborating-with-your-community-using-discussions/participating-in-a-discussion#marking-a-comment-as-an-answer)?

**Reply by werls** [reply]
Done!

**Reply by pdehaan** [reply]
Per https://www.11ty.dev/docs/config/#transforms, you can get the `inputPath` and `outputPath` via `this.inputPath` and `this.outputPath` respectively, assuming you're on the latest 1.0+ release.

```js
  // Eleventy 1.0+
  eleventyConfig.addTransform("transform-name", function (content) {
    console.log( this.inputPath );
    console.log( this.outputPath );

    return content; // no change done.
  });
```

Also, https://www.11ty.dev/docs/plugins/rss/#supplies-the-following-nunjucks-filters has a nice `absoluteUrl` filter which can be useful. I'd also probably say to specify that `https://my.website.url` in a global data file or somewhere a bit easier to manage if it's repeated in 1+ places (but that's more of a micro-nit thing).

I'd have to do some research into the [`@11ty/eleventy-img` plugin](https://www.11ty.dev/docs/plugins/image/#use-this-in-your-templates) to see how easy it'd be to add a custom class name to a specific image, which might also allow you to do something like this if you wanted to target specific images.

```js
$('main img.og-image');
```

Looking at https://ogp.me/#structured, it seems like there are a few other optional `og:image:*` properties you could set (like `type` for MIME types), or `width`/`height` for dimensions, or `alt` for adding a description of the image. Adding the `og:image:alt` might be trivial, given you're already grabbing the asset from the DOM using cheerio.

```js
$('head').append(`
  <meta name="og:image" content="https://my.website.url${$firstImage.attr('src')}" />
  <meta name="og:image:alt content="${$firstImage.attr('alt')}" />
`);
```

And like you said, if it's a portfolio site, there isn't much risk of a page not having an image, but you also might want to add a quick check around this to make sure that `$firstImage` isn't somehow null/undefined (if it was a contact/about page):
```js
const $firstImage = $('main img');
```
